### PR TITLE
Release 1.10.0

### LINKED ISSUE #3507: Memoize `getShopId` to Reduce Queries
### Expected behavior
Once the "current" shop id is determined, return a cached version to avoid querying the DB again.

### Actual behavior
On startup, we call `getShopId` more than 900 times
Per request, it's about 30.
(this method makes one or two calls to the database, depending on whether you are logged in and have an "active shop" selected).

### Steps to reproduce the behavior
place a `console.log` statement [here](https://github.com/pmn4/reaction/blob/master/server/api/core/core.js#L313) and [here](https://github.com/pmn4/reaction/blob/master/server/api/core/core.js#L322) to count the number of times we go to the database to get the current shop.

### Versions
master

### LINKED ISSUE #3866: Refactor Products.after.remove|update|insert collection hooks in search-mongo package.
## Prerequisites
* [x] Are you running the latest version?
* [x] Are you able to consistently reproduce the issue?
* [x] Did you search the issue queue for existing issue? 

## Issue Description
Remove `Products.before.remove|update|insert` collection hooks in mongo-search package, this is part of the effort to remove the collection hooks meteor package.

Hook call sites:
- https://github.com/reactioncommerce/reaction/blob/3634-willopez-remove-product-collection-hooks/imports/plugins/included/search-mongo/server/hooks/search.js#L56
- https://github.com/reactioncommerce/reaction/blob/3634-willopez-remove-product-collection-hooks/imports/plugins/included/search-mongo/server/hooks/search.js#L67
- https://github.com/reactioncommerce/reaction/blob/3634-willopez-remove-product-collection-hooks/imports/plugins/included/search-mongo/server/hooks/search.js#L56

**Umbrella Issue**: https://github.com/reactioncommerce/reaction/issues/3634 refer to this issue for more details. 
## Possible Solution
Refactor hook into a function and call explicitly where necessary. Refer to this PR for an example: https://github.com/reactioncommerce/reaction/pull/3825

### Versions
Node: 8.9.4
NPM: 5.6.0
Meteor Node: 8.9.4
Meteor NPM: 5.6.0
Reaction CLI: 0.28.0
Reaction: 1.8.0
Reaction branch: 3634-willopez-remove-product-collection-hooks
Docker: 17.12.0-ce

### LINKED ISSUE #3910: Replace GraphQL-Yoga deps with specific deps we use
**Summary**
We need to replace the more-lib-than-we-need GraphQL Yoga dependency with specific dependencies that we're actively using to implement GraphQL right now. We're not using Yoga right now and it's adding unnecessary weight.

## Acceptance Criteria
We will have removed the GraphQL Yoga dependency and replaced it with more specific dependencies that are actively used by Reaction.

### LINKED ISSUE #3928: Design and implement an ID system for use with GraphQL
**Summary**
Work with @ticean to design and implement an ID system for use with GraphQL queries and mutations.

**Implementation**
This system will likely involve prefixing/namespacing existing Ids and then Base64 encoding on the client. When going from the client to the server, it will involve Base64 decoding and removing of the prefix.

## Acceptance Criteria
This ticket will be resolved when we have a system for using IDs within GraphQL and there is an implemented abstraction that makes it simple to use.

### LINKED ISSUE #3935: Design error handling for GraphQL
* Decide how to return errors to the client in GraphQL queries, format etc.
* Decide how to handle errors in the implementation.

### LINKED ISSUE #3936: Create tests pattern for new GraphQL functions
Using one initial mutation and one resolver, create a branch with a proposed pattern for unit tests, preferably outside of Meteor as much as possible.

### LINKED ISSUE #3975: Outbound Shopify sync status not properly shown in dashboard
## Issue Description
If you enable "Push order to shopify on create" and select "Setup Hooks" the checkbox does not remain selected if you close the dashboard panel and return

## Steps to Reproduce
1. Login as admin
1. Go to Connectors dashboard panel
1. Click on "Push orders to Shopify on create"
1. Click on "Setup Hooks"
1. Close panel
1. Observe that checkbox is no longer selected

```
Node: 8.9.4
NPM: 5.6.0
Meteor Node: 8.9.4
Meteor NPM: 5.6.0
Reaction CLI: 0.28.0
Reaction: 1.9.0
Reaction branch: release-1.9.0
Docker: 17.12.0-ce
```

### LINKED ISSUE #3994: Remove Media hooks
Search for `MediaRecords.before.` in `/imports/plugins/core/revisions/server/hooks.js` and remove those hooks, but keep the revision control that they implement.

## For inserts

The only `Media.insert` that needs revision control is in /imports/plugins/core/ui/client/containers/mediaGallery.js

Move that insert to a Meteor method, and move the `MediaRecords.before.insert` logic into that method.

## For updates

Make an "updateMediaMetadata" method to replace `MediaRecords.before.update`.

Needs to be called in one place: /imports/plugins/core/files/server/methods.js in the `sortedMediaIDs.forEach`

## For removes

Make a "removeMedia" method to replace `MediaRecords.before.remove`.

Call "removeMedia" method instead of `Media.remove` directly in /imports/plugins/core/ui/client/containers/mediaGallery.js

Should be able to fix #3992 while doing this.

### PR DESCRIPTION
**Pull request by spencern**
# v1.10.0

## Collection Hooks | Breaking Change!
The biggest change in 1.10.0 is the removal of the [meteor-collection-hooks](https://github.com/matb33/meteor-collection-hooks) package. If your app or plugin is dependent on collection hooks, we'd recommend refactoring to eliminate the need for those types of hooks. If you're not up for refactoring you can just add that package back into your app though you'll lose some of the benefits that Reaction will see by eliminating that dependency. The removal of this package is a breaking change for any plugin that depends on collection hooks or uses the `direct` method to update a collection. For example, `Products.before`, `Products.after`, and `Products.direct` will no longer work. These hooks have all been migrated to our event hook system. If you were using Collection hooks to perform an action that you cannot perform in 1.10.0 with the current set of event hooks, please file an issue and we can determine if we need to add additional event hooks. This change should lead to much simpler code that is easier to understand.

## New GraphQL API
In addition to a few performance updates, refactors, and bug fixes, this release starts the rollout of our GraphQL API. Until further notice, this GraphQL API should be considered a prototype and should not be used for production. If you're interested in following along with our GraphQL work, tracking our [GraphQL issues](https://github.com/reactioncommerce/reaction/issues?utf8=%E2%9C%93&q=is%3Aissue+graphql+) is probably the easiest way for now.

In short, this release adds a prototype GraphQL server running in the Reaction project with the following urls exposed:

```
/graphql - The GraphQL endpoint.
/graphiql - The GraphiQL user interface.
```

We'll be incrementally adding functionality to this GraphQL endpoint, but this release serves as the starting point for our GraphQL api.

 ## GraphQL
 - feat: GraphQL Prototype (#3898) .. Resolves #3935, Resolves #3928, Resolves #3910
 - feat: GraphQL Jest testing pattern (#3995) .. Resolves #3936
 - feat: Create GraphQL viewer query (#4019)
 - feat: Create GraphQL account query (#3991)

## Refactors
 - refactor: remove collection hooks in search mongo package (#3889) .. Resolves #3866
 - refactor: remove collection hooks for Products collection (#3825)
 - refactor: remove media hooks (#4035) .. Resolves #3994
 - refactor: hooks in the inventory package to use Hooks.Events (#3887)
 - refactor: remove collection hooks package (#4036)

## Performance
 - perf: Memoize/Cache getShopId to Reduce DB Load (#3510) .. Resolves #3507

## Fixes
 - fix: typo fix (#4000) .. Resolves #3975
 - fix: Browser console warning when beginning checkout (#3980)
 - fix: PDP Image gallery does not handle portrait sized images well (#3993)
 - fix: Cloning products is not reactive (#3964)
 - fix: Uploading Product Image (#4029)
 - fix: ProductsContainers are not replaceable (#4025)
 - fix: ProductGridItem is not replaceable (#4027)

## Tests
 - CI: Run Jest and Meteor tests in parallel (#4030)

## Docs
 - docs(jsdoc): namespace MethodHooks into its own JSDoc section. (#3844) .. Resolves #3840

### CODE REVIEW

File: server/methods/core/cart.js [hunk 1]
```diff
@@ -101,6 +101,12 @@ function removeShippingAddresses(cart) {
   }, {
     $set: { shipping: cartShipping }
   });
+
+  // Calculate discounts
+  Hooks.Events.run("afterCartUpdateCalculateDiscount", cart._id);
```
Comment (zenweasel) [inline server/methods/core/cart.js]: Also here, why do we tie this to specific events, what if I want to do something else besides calculate discounts and taxes?

**Comment by zenweasel** [comment]
Not a critical bug but broken in this release: https://github.com/reactioncommerce/reaction/issues/4017

**Comment by zenweasel** [comment]
Blocked by #4018 but (probably) same cause as #4016

**Comment by lcampanis** [comment]
@spencern  is this release still considered unstable for Production?

**Comment by zenweasel** [comment]
@lcampanis This release is basically unusable at this point. Hold off for a bit

**Comment by spencern** [comment]
@lcampanis once this release is out, it will be considered stable, but my notes about the GraphQL API stand and should not be used for production even once this release is out. While we're rolling out GraphQL along side our stable branch, it's subject to change at any point.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,51 @@
+# v1.10.0
+
+## Collection Hooks | Breaking Change!
+The biggest change in 1.10.0 is the removal of the [meteor-collection-hooks](https://github.com/matb33/meteor-collection-hooks) package. If you're app or plugin is dependent on collection hooks, we'd recommend refactoring to eliminate the need for those types of hooks. If you're not up for refactoring you can just add that package back into your app though you'll lose some of the benefits that Reaction will see by eliminating that dependency. The removal of this package is a breaking change for any plugin that depends on collection hooks, or uses the `direct` method to update a collection. For example, `Products.before`, `Products.after`, and `Products.direct` will no longer work. These hooks have all been migrated to our event hook system. If you were using Collection hooks to perform an action that you cannot peform in 1.10.0 with the current set of event hooks, please file and issue and we can determine if we need to add additional event hooks. This change should lead to much simpler code that is easier to understand.
```
Comment (zenweasel) [inline CHANGELOG.md]: *your

Comment (spencern) [reply CHANGELOG.md]: good catch.

**Pull request merged**